Langevin PR

andyElking opened the pull request:
Hi Patrick,

Thanks for the heads up, here's the reuploaded PR (with another quick fix).

This PR contains all the new Langevin solvers. All of these inherit from AbstractLangevinSRK in langevin_srk.py.
Another important addition is LangevinTerm in _term.py. I explained why it is needed in a comment bellow.

I haven't added the new solvers to the docs and autocite yet, because 1) the relevant paper is not on arxiv yet, but might be in a month or two and 2) I expect you might suggest several changes, so might as well write the docs once the rest is stationary. Still, I think the docstrings and comments are quite comprehensive.

I'm making this PR now so you have ample time to have a look at it, but I will be away for the next few weeks, so there is aboslutely no hurry.

Best,
Andraž

andyElking commented on diffrax/_term.py:
```diff
@@ -776,3 +785,135 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    return jtu.tree_map(_inner_broadcast, source, target_tree_shape)
+
+
+class LangevinTerm(AbstractTerm):
```
I introduced `LangevinTerm` for the reasons below, but I am happy to change this if you think there's a better way to handle this.
The requirements are:
(1) this term can be passed into any of the regular SRKs, which only accept a
MultiTerm[ODETerm, AbstractTerm]
(2) this term needs to carry an `args` attribute. You said that
since ALIGN, SORT and ShOULD, use the `args` in a very special way, we
cannot just pass them as usual (i.e. in `vf(t, y, args)`), but that the args
should be part of the term structure itself.

Frankly, I could just use a tuple (args, MultiTerm[ODETerm, AbstractTerm])
and pass that around, but the issue is that if the user mishandles
the tuple (e.g. changes the args but not the term), then the solution from
ALIGN (which depends on the args) will be different than the one from SRK,
which depends only on the MultiTerm (which has the args baked into the VF).

In a way `LangevinTerm` is just a wrapper for a `MultiTerm`.

patrick-kidger replied:
I think having this as a separate term seems totally reasonable to me.

patrick-kidger replied:
Although, looking at what you have here: this is basically a custom constructor for a `MultiTerm`, right?

In that case why not just make this a function, rather than a class, which returns a `MultiTerm` with the appropriate inputs?

andyElking replied:
No, that's my whole point: it's not just a custom `MultiTerm` constructor. It is a container that holds both the arguments `(gamma, u, grad_f)` and the `MultiTerm`. The latter can be computed from the former, but not the other way round (at least not in a simple way). And `AbtractLangevinSRK` solvers need just the arguments, whereas all other solvers need just the `MultiTerm`. Hence both of these objects need to be carried around and then when `diffeqsolve` is called, depending on the solver, it either passes the `MultiTerm` or the arguments. I am very strongly convinced that the correct behaviour here cannot be achieved without both modifying `diffeqsolve` and introducing a new term. Maybe we should have a call to discuss this.

andyElking commented:
Hi @patrick-kidger, just bumping this in case you didn't notice that the tests passed. No hurry though :)

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -123,6 +123,9 @@ def _term_compatible(
     term_structure: PyTree,
     contr_kwargs: PyTree[dict],
 ) -> bool:
+    if term_structure == LangevinTerm and isinstance(terms, LangevinTerm):
+        return True
```
I don't think this kind of special-casing is okay. Defining custom terms like `LangevinTerm` should be a thing that users can do just as well, it shoudn't require special support in the core of Diffrax.

andyElking replied:
Without this it throws and error. I don't know how to work around this. The issue that Langevin term is trying to solve is quite convoluted. I'll try to explain it in a way that also answers some of your later comments.

So `LangevinTerm` is not just a custom constructor for a `MultiTerm`. The issue is that any of the `AbstractLangevinSRK` solvers only need the term to provide them with the arguments `(gamma, u, grad_f)`, which uniquely determine the Underdamped Langevin Diffusion (ULD). However, all other non-Langevin solvers need to be provided with just a `MultiTerm` which defines that particular ULD. But given the vector fields provided by the `MultiTerm`, it is impossible to deduce the arguments `gamma, u, grad_f`, so we need some object which carries both these arguments and the `MultiTerm` derived from these arguments (and the fact that it is a ULD).

This means that `AbstractLangevinSRK` needs to have the term structure `LangevinTerm` (and can accept no other term), but `_term_compatible` simply fails when I do `diffeqsolve(..., ALIGN(), LangevinTerm(...), ...)`. I think the way `_term_compatible` is written right now, it is impossible for solvers to have term structures which contain custom terms. So yes, this special-casing is not okay, but `_term_compatible` as it is written now does not work as intended.

patrick-kidger replied:
Ah, I think I see what you need here is to actually structure things as:
```python
MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]
```
where the drift/diffusion are parameterised by `(gamma, u, grad_f)`

The Langevin-specific solvers can then be written to accept precisely `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]` as their `term_structure`. Meanwhile we'll be still be compatible with the other general SDE solvers because we're already a `MultiTerm` of the right form.

(Actually I think technically we might need to generalise some of the SDE solvers from accepting `MultiTerm[ODETerm, AbstractTerm]` into accepting `MultiTerm[AbstractTerm[RealScalarLike], AbstractTerm]`, but that should just be a one-line change.)

andyElking replied:
Nevermind, I found the issue:
until now `LangevinTerm.term` was not a static field, so `_term_compatible` went digging down into the PyTree instead of just treating the `LangevinTerm` as a `LangevinTerm`. All it took was to add a ` = eqx.field(static=True)`. My bad, sorry for the whole confusion.

Still, I hope it is fine if I keep the following in `_integrate.py`:
https://github.com/patrick-kidger/diffrax/blob/15c21a66206690dde470cb0d4f77b25471758d32/diffrax/_integrate.py#L1011-L1013

If not, then I can probably try doing as you suggest and change it all into a `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]`. I feel like this is the sort of scenario where inheriting from a concrete class (i.e. `LangevinTerm` inheriting from `MultiTerm`) would make things much simpler :). Wrappers, however, need to be unwrapped, requiring changes in other parts of the codebase. (I do agree though that the pros of "concrete means final" still outweigh the cons).

patrick-kidger replied:
I definitely think we should avoid special-casing this one term in `_integrate.py`. The terms we ship with shouldn't really be special, and users should be able to implement their own in the same way as us. (In some sense the only two 'special' terms we have are `AbstractTerm` and `MultiTerm`.)

Do you forsee something potentially going wrong with the `MultITerm[LangevinDriftTerm, LangevinDiffusionTerm]` approach?

andyElking replied:
Yes, there are a few things that would go wrong with `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]`. Can we just have a call? I feel like we're spinning in circles.

andyElking replied:
I did more digging and the thing that goes wrong with the current situation is the following:
```
diffrax/_integrate.py:170: in _check
    vf_type = eqx.filter_eval_shape(term.vf, 0.0, yi, args)
venv/lib/python3.11/site-packages/equinox/_eval_shape.py:38: in filter_eval_shape
    dynamic_out, static_out = jax.eval_shape(ft.partial(_fn, static), dynamic)
venv/lib/python3.11/site-packages/jax/_src/api.py:2808: in eval_shape
    return jit(fun).eval_shape(*args, **kwargs)
venv/lib/python3.11/site-packages/jax/_src/pjit.py:494: in eval_shape
    _, params, _, _, out_tree, _, _, _, _ = _infer_params(jit_info, args, kwargs)
venv/lib/python3.11/site-packages/jax/_src/pjit.py:622: in _infer_params
    in_shardings_flat, in_layouts_flat = _process_in_axis_resources(
venv/lib/python3.11/site-packages/jax/_src/util.py:313: in wrapper
    return cached(_IgnoreKey(), *args, **kwargs)
<string>:4: in __eq__
    ???
venv/lib/python3.11/site-packages/jax/_src/core.py:753: in __bool__
    check_bool_conversion(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

arr = Traced<ShapedArray(bool[2])>with<DynamicJaxprTrace(level=1/0)>

    def check_bool_conversion(arr: Array):
      if arr.size == 0:
        raise ValueError("The truth value of an empty array is ambiguous. Use"
                         " `array.size > 0` to check that an array is not empty.")
      if arr.size > 1:
>       raise ValueError("The truth value of an array with more than one element"
                         " is ambiguous. Use a.any() or a.all()")
E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

venv/lib/python3.11/site-packages/jax/_src/core.py:654: ValueError
```

Given how mysterious `eqx.filter_eval_shape` is to me, I have no way of debugging this further.

lockwo replied:
I was playing around with this, and I noticed a couple things. The error only showed up if I ran more than one test. I was narrowing it down and commented everything else but one test and it worked, but 2 tests (of any kind) errored. 

My current hypothesis is that something weird is going on with the static field. I'm not 100% up to date on equinox's static field, but I believe they get cast as aux data in the module pytree and jax requires aux data to be hashable (https://stackoverflow.com/questions/78485445/why-is-custom-pytree-aux-data-traced-after-jax-jit-for-jnp-array-but-not-for). And you have a tuple with array values (the hash of which depends on the elements https://stackoverflow.com/questions/49722196/how-does-python-compute-the-hash-of-a-tuple). It seems weird that if this were true, that it wouldn't immediately error with one, but I have had weird edge case failures with hash + jax array. However, changing the args to be static, but be reliably hashable fixed it for me (see https://github.com/lockwo/diffrax/tree/langevin_test, just one commit ahead of your branch) which I why I think this is it. I don't think my solution I did there is really great, but I think this is basically an edge case of being the only term class that has array elements, so it might require some rethinking/generalization of the term class interactions (maybe as simple as adjusting some tree maps to ignore anything that isn't a specific term?). 

Also, I think I get why the Langevin term was introduced, but my question is, rather than having a MultiTerm of drift and diffusion, would making an abstract multiterm then having Langevin term inherit from that make sense? Like it is (conceptually) a multiterm, so there could be the other multiterm and the Langevinmultiterm. Just a thought I had while experimenting.

andyElking replied:
Thanks Owen, this is really useful information! Patrick and I are having a call this Friday at 12 noon PST (8pm UK time). If you have time to join maybe @patrick-kidger could add you to the call?

patrick-kidger commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+    taylor_threshold: RealScalarLike = eqx.field(static=True)
+    _coeffs_structure = jtu.tree_structure(
+        _ALIGNCoeffs(
+            beta=jnp.array(0.0),
+            a1=jnp.array(0.0),
+            b1=jnp.array(0.0),
+            aa=jnp.array(0.0),
+            chh=jnp.array(0.0),
```
In some cases it's not valid to use JAX arrays at the global level. Better to do what we do in `AbstractRungeKutta` and instead use numpy arrays here, and then convert them during the forward pass.

At least in this case I can see you're just using it for the structure, in which case just a `0` or similar will suffice. (Tweaking type hints or something if required.)

andyElking replied:
Honestly, I didn't want to change the type annotations just because of this dummy structure-helper. But I guess I'll change it to `ArrayLike`.

patrick-kidger commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+        )
+    )
+
+    @property
+    def minimal_levy_area(self):
+        return AbstractSpaceTimeLevyArea
```
I don't think this needs to be a property?

andyElking replied:
Yes, of course it should just be a field.

patrick-kidger commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+    def minimal_levy_area(self):
+        return AbstractSpaceTimeLevyArea
+
+    def __init__(self, taylor_threshold: RealScalarLike = 0.0):
+        r"""**Arguments:**
+
+        - `taylor_threshold`: If the product `h*gamma` is less than this, then
+        the Taylor expansion will be used to compute the coefficients.
+        Otherwise they will be computed directly. When using float32, the
+        empirically optimal value is 0.1, and for float64 about 0.01.
+        """
+        self.taylor_threshold = taylor_threshold
```
Maybe let's set this to `0.1` by default?

Also it looks like you could assign this default value directly to the field and skip having an `__init__` method here?

andyElking replied:
Yeah, I originally had it at 0.1 by default but then Trevor asked me to change it to 0. But I agree with you (and hopefully Trevor won't notice :)).

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+_Coeffs = TypeVar("_Coeffs", bound=_AbstractCoeffs)
+
+
+# TODO: I'm not sure if I can use the _Coeffs type here,
+# given that I do not use Generic[_Coeffs] in the class definition.
+# How should I work around this?
```
What goes wrong with using `Generic` in the class definition?

andyElking replied:
Hmm, I remember pyright complaining last time, but now I tried adding the `Generic` again and it works, so good I guess.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+    _coeffs_structure: eqx.AbstractClassVar[PyTreeDef]
+    minimal_levy_area: eqx.AbstractClassVar[type[AbstractBrownianIncrement]]
+
+    @staticmethod
```
I don't think these need to be static. In principle a concrete class may want something that is `self`-dependent?

(In practice the ones you have right now might not want this, in which case you can just do `del self` inside the body of the method.)

andyElking replied:
Fair enough. But knowing the code, I doubt a concrete class would ever need something `self`-dependent here.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+# one can compute the SRK coefficients more precisely for a small h.
+
+
+class AbstractLangevinSRK(
+    AbstractStratonovichSolver[_SolverState],
+    AbstractItoSolver[_SolverState],
+    Generic[_Coeffs, _ErrorEstimate],
+):
```
Can you add a docstring here?

FWIW if this is based on James' theory then I suspect we should use a more precise name than `AbstractLangevinSRK`. Call it `AbstractFosterLangevinSRK` or similar.

andyElking replied:
Yes, I added a short docstring.

I aggre and I can try to convince him, but I doubt he will agree to this naming scheme. (we could do it without him knowing though :))

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+
+    @staticmethod
+    @abc.abstractmethod
+    def _tay_cfs_single(c: Array) -> _Coeffs:
```
I think you're using both `coeffs` and `cfs` as an abbreviation? Let's go with just `coeffs`.

andyElking replied:
Good point

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+    @staticmethod
+    def _eval_taylor(h, tay_cfs: _Coeffs) -> _Coeffs:
+        # Multiplies the pre-computed Taylor coefficients by powers of h.
+        # jax.debug.print("eval taylor for h = {h}", h=h)
```
?

andyElking replied:
Oops

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+
+                    return jtu.tree_map(_choose, tay_out, direct_out)
+
+                # If all entries of h*gamma are below threshold, only compute tay_out
+                # otherwise, compute both tay_out and direct_out and select the
+                # correct one for each dimension
+                return lax.cond(
+                    jnp.all(cond), lambda _: tay_out, select_tay_or_direct, None
+                )
```
Do you have any opinion on how this should behave around batching?

If the predicate of a `lax.cond` is batched then both branches are computed unconditionally and then `jnp.where`'d together.

If you're doing this logic here just for efficiency, then you might want to swap out the predicate for something like `eqxi.unvmap_all`, whilst taking care to ensure that the internal logic of both branches remains valid.

andyElking replied:
Yes, um I agree this should be improved. I'll keep it pending for now and address it in alater commit.

andyElking replied:
I thought this would require more work, but yeah simply substituting in `eqxi.unvmap_all` worked like a charm. And indeed both branches are always logically valid, that's for sure. This is a very useful trick, thanks for recommending!

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+        cfs: _Coeffs = st.coeffs  # type: ignore
+
+        # If h changed recompute coefficients
+        cond = jnp.isclose(h_state, h)
```
I'd suggest perhaps making this equality? This seems like we're leaving numerical correctness kind of up to chance here.

andyElking replied:
Maybe I can make the tolerance smaller, but for some reason requiring equality would make the coeffs be recomputed way too often (although it's been a while since I experimented with this, so I could be wrong).

andyElking replied:
So yes I redid the experiments and even with a constant step controller, the step sizes fluctuate by cca 1e-16, which is negligible, but if I have `==` instead of `jnp.isclose`, then the coefficients get recomputed way too often. So I'll stick with `cond = jnp.isclose(h_state, h, rtol=1e-10, atol=1e-12)`.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,289 @@
+
+        # If h changed recompute coefficients
+        cond = jnp.isclose(h_state, h)
+        cfs = lax.cond(
```
Same comment as before on `cfs` vs `coeffs`.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,135 @@ def _to_vjp(_y, _diff_args, _diff_term):
         return dy, da_y, da_diff_args, da_diff_term
+
+
+LangevinX = PyTree[Shaped[Array, "?*langevin"], "LangevinX"]
```
It's not clear to me from the name what this quantity represents.

andyElking replied:
I agree it is unclear, I added the following comment:

```
# The Underdamped Langevin SDE trajectory consists of two components: the position
# `x` and the velocity `v`. Both of these have the same shape. So, by LangevinX we
# denote the shape of the x component, and by LangevinTuple we denote the shape of
# the tuple (x, v).
```

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,135 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    def __init__(self, args, bm: AbstractBrownianPath, x0: LangevinX):
+        r"""**Arguments:**
+
+        - `args`: a tuple of the form $(\gamma, \, u, \, \nabla \! f)$
```
Why are these together in `args`, rather than separate arguments?

patrick-kidger replied:
Also, I think spacing commands like `\!` should be avoided in LaTeX like this, as (a) it's often read directly, and (b) many renderers don't support them.

andyElking replied:
> Why are these together in `args`, rather than separate arguments?

Because it makes it easier to pass them on that way. I swapped to have them separate at some point, but I think it looked more awkward.

> Also, I think spacing commands like `\!` should be avoided in LaTeX like this, as (a) it's often read directly, and (b) many renderers don't support them.

Fair enough, I'll get rid of it.

patrick-kidger reviewed (commented):
Haha, sorry, it's taken me a while to get to this!

Here's a first review. Many of my comments on `align` apply to the others too, I think.

andyElking commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+
+
+class _ALIGNCoeffs(_AbstractCoeffs):
+    @property
+    def dtype(self):
+        return jtu.tree_leaves(self.beta)[0].dtype
```
On one hand I do know they will all have the same dtype. But yes, I'll just make a field that gets computed at init.

andyElking commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+    """
+
+    term_structure = LangevinTerm
+    interpolation_cls = LocalLinearInterpolation
```
I asked him, but it might not be possible.

andyElking replied:
Update: James found an interpolation scheme for ALIGN, which I will implement and test when I have the time. However, he thinks it is very unlikely we'd be able to find any interpolation schemes for third-order methods, which would be any better than piecewise-linear. Is it okay if we finish this PR first and I add the ALIGN interpolation afterwards?

patrick-kidger replied:
Yup, we can definitely do this in a later PR.

andyElking commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+    def dtype(self):
+        return jtu.tree_leaves(self.beta)[0].dtype
+
+    beta: PyTree[Array]
+    a1: PyTree[Array]
+    b1: PyTree[Array]
+    aa: PyTree[Array]
+    chh: PyTree[Array]
```
Yes, but it's very very awkward. I tried using shape constraints but in the end it was just useless nuissance. Also these coefficients are only used internally and I use asserts at appropriate points in the code to check they all have the correct shapes. I am really very certain there is no point in adding constraints here.

andyElking commented on diffrax/_solver/align.py:
```diff
@@ -0,0 +1,164 @@
+            aa=aa,
+            chh=chh,
+        )
+        return jtu.tree_map(lambda x: jnp.array(x, dtype=dtype), out)
```
Actually I think this is entirely superfluous, so I got rid of it.

andyElking commented:
Hi Patrick,

Thanks so much for your review! I addressed almost all of your comments. Two of them I will address in a later commit (it's getting late today haha).

I think you didn't quite understand the reason why the `LangevinTerm` and the changes in `_integrate.py` are needed, so I tried to give an explanation, but if that is unclear we can have a call at some point. Let me know :)

andyElking commented:
Quick heads up: I now made all the edits you suggested and the tests all passed :)

andyElking commented:
@patrick-kidger a note about interpolation for ALIGN:
to maintain a 2nd order of convergence at interpolated points, the interpolated value cannot depend just on `t0, t, t1, y0, y1`, but must also depend on `W0, H0, W1, H1`. Do you think it would be feasible to modify the interpolation code in order to allow for that? Maybe I could store W and H as part of the solution? Do you have any ideas how to do this in a way that fits into Diffrax naturally?

patrick-kidger commented:
> to maintain a 2nd order of convergence at interpolated points, the interpolated value cannot depend just on t0, t, t1, y0, y1, but must also depend on W0, H0, W1, H1. Do you think it would be feasible to modify the interpolation code in order to allow for that? Maybe I could store W and H as part of the solution? Do you have any ideas how to do this in a way that fits into Diffrax naturally?

I think should be totally fine -- it can go into the `dense_info` output of a solver step. Take a look at how Runge--Kutta methods output the intermediate stages, for example.

andyElking commented:
@patrick-kidger I now made a temporary fix, but there are two things that remain to be solved:

1. For some reason `eqx.filter_eval_shape(term.vf, 0.0, y0, args)` doesn't work when `term` is a `LangevinTerm` (the stacktrace is in a comment above). I genuinely do not undertand what is going wrong, so please help.
2. I think we are still not on the same page as to why I introduced `LangevinTerm` in the first place. Please let's have a call at some point to discuss it. And then hopefully it will make sense why using `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]` could lead to incorrect results.

andyElking commented:
Great news @patrick-kidger @lockwo: what we discussed worked! Thanks for your advice!

Patrick, if there is nothing else about the code itself that you'd like me to change, then I'll add all of the new things to the docs. Do you think I should add a short example of how to use the langevin solvers as well? Maybe a simple Langevin Monte Carlo example?

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
+# the tuple (x, v).
+_LangevinX = PyTree[Shaped[Array, "?*langevin"], "LangevinX"]
+_LangevinTuple = tuple[_LangevinX, _LangevinX]
+_LangevinBM = TypeVar("_LangevinBM", bound=Union[_LangevinX, AbstractBrownianIncrement])
```
I don't understand this type variable -- it seems to go unused below. (I.e. I don't see `_LangevinDiffusionTerm[...]` anywhere.)

I think you probably just want to inline this `Union` into `_LangevinDiffusionTerm(AbstractTerm[...]):`?

andyElking replied:
Good point! I think this is a remnant of when I was first coding this and Python's type system felt pretty foreign to me (it still does sometimes).

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> _LangevinBM:
+        return self.control.evaluate(t0, t1, **kwargs)
+
+    def prod(self, vf: _LangevinX, control: _LangevinX) -> _LangevinTuple:
+        if isinstance(control, AbstractBrownianIncrement):
+            control = control.W
```
This type annotation for `control` would imply that the `isinstance` check never triggers.

(I'm surprised some of the runtime typechecking during tests doesn't pick up on this -- if you ever pass an `AbstractBrownianIncrement` then I think it should fail to match the structure of `_LangevinX` as defined by the `vf` term.)

Morally speaking should we have this `isinstance` check, do you think? I note that we don't have it on any of the other terms -- we ask that the solvers handle this for us, rather than the term.
(No strong feelings, I'm not sure what's cleanest.)

andyElking replied:
You're very correct! Honestly I don't remember why I put that there. Indeed I elsewhere always stick to the rule that `prod` should never be fed an `AbstractBrownianIncrement`, since unpacking that should be done by the solvers, as you say.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    bm: AbstractBrownianPath,
+    x0: _LangevinX,
+) -> MultiTerm[tuple[_LangevinDriftTerm, _LangevinDiffusionTerm]]:
+    r"""Creates a term that represents the Underdamped Langevin Diffusion, given by:
+    $d \mathbf{x}_t = \mathbf{v}_t dt$
+    $d \mathbf{v}_t = - \gamma \mathbf{v}_t dt - u
+    \nabla f( \mathbf{x}_t ) dt + \sqrt{2 \gamma u} d W_t.$
+    where $\mathbf{x}_t, \mathbf{v}_t \in \mathbb{R}^d$ represent the position
+    and velocity, $W$ is a Brownian motion in $\mathbb{R}^d$,
+    $f: \mathbb{R}^d \rightarrow \mathbb{R}$ is a potential function, and
+    $ \gamma,u\in\mathbb{R}^{d\times d}$ are diagonal matrices governing
+    the friction and the dampening of the system.
```
Can you check whether these render correctly in the docs? (I suspect you'll need to a couple of newlines and things.)

andyElking replied:
You're right, thanks!

Also I have an issue with my docs. For some reason the class headers are not appearing. So where you have `::: diffrax.ODETerm` the docs don't display "ODETerm" but just its docstring. Do you know how to fix this? I googled around but couldn't find anything.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
+        This callable should take a PyTree of the same shape as `x0` and
+        return a PyTree of the same shape.
+    - `bm`: A Brownian path representing the Brownian motion $W$.
+    - `x0`: The initial state of the system (just the position without velocity);
+        only needed to check the PyTree structure of the other arguments.
```
Is this argument really necessary?

Note that elsewhere in Diffrax we don't take `y0` as an input to our terms -- rather, we allow our term definitions to be polymorphic over many possible `y0`s. E.g. `ODETerm(lambda t, y, args: -y)` does not say anything about the shape of `y`.

I suspect the same should probably be doable here by moving the broadcasting into the new terms themselves.

andyElking replied:
I rely on gamma to be the right shape for the coefficients in `LangevinSRK`. So there are several options I see:
- recast `gamma` and `u` into the correct shapes in `LangevinSRK.init` and hold them in solver state (honestly might not be so bad)
- assume that `gamma` will have the same shape as `x0` (which I don't want to require, since it should be fine to have a scalar `gamma` even if `x` is a weird PyTree)
- do what is currently done.

To me the last option seems a bit more robust a bit more transparent to the user: if their arguments are of incorrect shapes, then a Langevin term cannot be made in the first place. Whereas if a solver complains about some PyTree mismatch, then it might not be as obvious what went wrong (or at least I'd have to fiddle with `try, catch` to provide clear error messages, etc). And I doubt this hinders the usability, since they'll need to have an `x0` somewhere anyway. But I am happy to go with the first option.

patrick-kidger replied:
I think the first option is probably my favourite! I'd like to keep this abstraction the way it is.

(The reason it is the way it is, by the way, is because of cases like https://docs.kidger.site/diffrax/examples/neural_ode/ -- we often want to be able to describe the vector field at `__init__` time whilst only obtaining the initial condition at `__call__` time.)

andyElking replied:
Sounds good!

patrick-kidger commented on test/helpers.py:
```diff
@@ -176,7 +181,10 @@ def _batch_sde_solve(
     )
-    return sol.ys, sol.stats["num_accepted_steps"]
+    steps = sol.stats["num_accepted_steps"]
+    if isinstance(solver, diffrax.HalfSolver):
+        steps *= 3
```
I don't think this is fair! Different solvers anyway have different per-step costs; if the desire is to normalize over that then that should be done elsewhere across all kinds of solvers.

andyElking replied:
You're right this doesn't have to be present in this branch.

patrick-kidger commented on test/helpers.py:
```diff
@@ -281,26 +296,40 @@ def sde_solver_strong_order(
 
 
-@dataclasses.dataclass(frozen=True)
+@dataclasses.dataclass(unsafe_hash=True)
```
Eek, it's unsafe for a reason. Why?

andyElking replied:
Hmm I remember there being a problem with frozen, but not sure what. I'll change it back and then if all hell breaks loose we'll know why.

patrick-kidger commented on test/helpers.py:
```diff
@@ -281,26 +296,40 @@ def sde_solver_strong_order(
+
+    @property
+    def w_shape(self):
+        if is_tuple_of_ints(self._w_shape):
+            return self._w_shape
+        else:
+            return self._w_shape
```
?

andyElking replied:
A complete mystery to me as well. I did get a good laugh out of it though. And why I needed it to be a property is also beyond me.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+import jax.tree_util as jtu
+from equinox import AbstractVar
+from jax import vmap
+from jax._src.tree_util import PyTreeDef
```
I don't think this requires a `_src` import.

(If nothing else it's available in `jaxtyping`.)

andyElking replied:
This one is on PyCharm (and me not checking 😅)

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+import abc
+from typing import Any, Callable, Generic, TypeVar
```
Prefer `from collections.abc import Callable`. The `typing` one is now deprecated.

andyElking replied:
Good to know, thanks!

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import (
+    _LangevinDiffusionTerm,
+    _LangevinDriftTerm,
+    _LangevinTuple,
+    _LangevinX,
```
If we're going to import these then we should make them public.

andyElking replied:
Makes sense. I assume you mean just removing underscores, not adding them to `__init__.py`?

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+
+
+def get_args_from_terms(
+    terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
```
I think this type annotation (and others below) are too broad -- it should accept specifically Langevin drift/diffusions.

andyElking replied:
What about `WrapTerm`s?

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+        unwrapped_drift = drift.term
+        assert isinstance(diffusion, WrapTerm)
+        unwrapped_diffusion = diffusion.term
+        assert isinstance(unwrapped_drift, _LangevinDriftTerm)
+        assert isinstance(unwrapped_diffusion, _LangevinDiffusionTerm)
+        gamma = unwrapped_drift.gamma
+        u = unwrapped_drift.u
+        f = unwrapped_drift.grad_f
+    else:
+        assert isinstance(drift, _LangevinDriftTerm)
+        assert isinstance(diffusion, _LangevinDiffusionTerm)
+        gamma = drift.gamma
+        u = drift.u
+        f = drift.grad_f
```
I think you can merge these two pieces. (Just set `drift = drift.term` and `diffusion = diffusion.term` on the `WrapTerm` branch.)

Will this do the right thing when solved backward in time?

andyElking replied:
Yes right the two branches were separate because I previously didn't have this as a separate function and I wanted to keep hold on the "wrapped" versions, precisely for the sake of direction. Because I use `term.contr` both for `h` and `levy` I am pretty convinced it should do the right thing. I'll write a test tomorrow.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+_Coeffs = TypeVar("_Coeffs", bound=AbstractCoeffs)
+
+
+# TODO: I'm not sure if I can use the _Coeffs type here,
+# given that I do not use Generic[_Coeffs] in the class definition.
+# How should I work around this?
```
This is old and can be removed I think?

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,325 @@
+    where $v$ is the velocity, $f$ is the potential, $gamma$ and $u$ are the
+    friction and momentum parameters, and $W$ is a Brownian motion.
+
+    Solvers which inherit from this class include ALIGN, SORT, ShOULD, and
+    QUIC_SORT.
```
Can you turn these into links in the documentation?

This can be done by styling them as e.g. ``[`diffrax.ALIGN`][]``

patrick-kidger reviewed (commented):
Nice! I really like this approach -- I think this is now looking really elegant.

I've gone through and left comments but they're mostly nits and tidy ups. I think we've now cracked the structure of this problem! :)

patrick-kidger commented:
> Do you think I should add a short example of how to use the langevin solvers as well? Maybe a simple Langevin Monte Carlo example?

I like the sound of that! An example would be great.

I'll emphasise 'short' -- I really try to keep the examples pedagogical.

andyElking commented:
Thanks so much for the review, Patrick! And sorry for all my dummy mistakes 😅. I made all the smaller edits already and tomorrow I'll write a short example, a test for the backward solve and put everything into the docs.

andyElking commented:
I added a brief Langevin example, a backwards in time test and fixed up the docs. I also did another check of the code, so hopefully there's no more rogue comments lurking in there. I also made sure all the docs render correctly.

lockwo commented:
I don't want to violate the "short" directive, but I think the Langevin example could be good with a little more meat (depending on which sort of audience/usage it is catered for). E.g. to target the sampling/ML people a little more, maybe adding one more plot showing autocorrelation time/ESS of ULD specific solvers vs EM on the HO problem. Or for physics community you could just add a plot showing some convergence as a function of friction. Just my opinion, since its mostly a tutorial on "how do I use ULD solvers" which seems like docs whereas examples are like "what is interesting/what can I uniquely do with these solvers".

andyElking commented:
That's a good point, I wanted to keep it short due to what Patrick said, but if he agrees I'm happy to make it longer. Also it seems you have some good ideas, so I certainly wouldn't mind if you wrote up something like that 😁.
I do have some notebooks lying around where I use it  for Bayesian logistic regression, but that is very far from short 😅.

lockwo commented:
Sure I can write something up that I think maintains the spirit of the "short" directive while adding a little extra spice to get people excited by these solvers

lockwo commented:
That's a good example. I got some good examples working, but they also rely on more complex neural network energy functions (which just adds up in terms of LoC), at least adding a little more pretty pictures: https://github.com/andyElking/diffrax_STLA/pull/4.

lockwo commented:
I'm actually revisiting your paper (https://arxiv.org/pdf/2405.06464), and I'm curious how it can have such a high strong order but have basically the same ESS as constant Euler? Is ESS/sampling just a problem that doesn't require convergence in a strong sense (seems like path wise convergence might matter for autocorrelations but idk)? Did over damped Euler (i.e. traditional Langevin MCMC) also perform comparably?

andyElking commented:
Two things to mention:

- Even the Euler discretisation of ULD has negative Lyapunov exponents (i.e. two paths with different initial conditions but same BM will converge together over time) as long as it doesn't explode. In other words, while the solution is stable, the importance of the initial condition (or past errors) vanishes exponentially.
- Other than that, if you keep adding sufficiently large independent Gaussians (or other sufficiently nice RVs), then the autocorrelation will go down fast. And as you can see Euler (mistakenly) explores a lot of space (due to instability), so its ESS is actually too big.

So you can converge to the wrong distribution and have a large ESS. Or you can converge to the correct distribution (like anything Metropolis adjusted should, in theory), but have a very low ESS and thus take forever to converge. The remarkable thing about SORT, however is that it has both a high ESS AND the correct distribution.

LMK if that answers your question.

lockwo commented:
"In other words, while the solution is stable, the importance of the initial condition (or past errors) vanishes exponentially." This is generally true of samplers right? Like exponential convergence away from initial condition, at least for discrete samplers, is necessary (since it's just the second largest eigenvalue of the transition matrix). I assume the same is true for continuous samplers.

"So you can converge to the wrong distribution and have a large ESS" I guess this is my confusion. Is this just because they don't meet the requirements of usual samplers. That is to say, any (valid, ergodic, detailed balance fulfilling, etc etc) sampler (e.g. HMC, MALA, RWMH, etc.) will necessarily converge to the true distribution (in infinite time). So any sample you draw after waiting equilibration time (characterized by the eigenvalues of the Fokker Planck operator) is a sample from the true distribution. And ESS dictates the amount of time I have to wait before drawing another sample (to erase the autocorellation), which seems good. So I guess my question is, is the problem that like ULA, Euler (and I guess QUICSORT) ULD are biased samplers (that do not meet my assumption above) because we don't have decreasing stepwise and we don't have any metropolis step/hastings correction? Or is your point just about the first step/getting to the true distribution (and measuring ESS before reaching it)?

andyElking commented:
Let's continue over email.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+
+        tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, gamma, tay_coeffs)
+        outer = jtu.tree_structure(gamma)
+        inner = self._coeffs_structure
```
I don't think you actually need `self._coeffs_structure` here. You should be able to determine this automatically from what is already here.

Probably simplest is something like
```python
inner = sentinel = object()

def recompute_coeffs_leaf(...):
    out = ...
    nonlocal inner
    if inner is sentinel:
        inner = jtu.tree_structure(out)
    else:
        assert inner == jtu.tree_structure(out)
    return out

tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, ...)
coeffs_with_tree = ...  # as before
```

andyElking replied:
That's such a great trick! Thank you!

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+        )
+
+    def _recompute_coeffs(
+        self, h, gamma: LangevinX, tay_coeffs: PyTree[_Coeffs], state_h
```
I think `state_h` is unused?

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+
+        return state_out
+
+    @staticmethod
+    @abc.abstractmethod
+    def _compute_step(
+        h: RealScalarLike,
+        levy,
+        x0: LangevinX,
+        v0: LangevinX,
+        langevin_args: _LangevinArgs,
+        coeffs: _Coeffs,
+        st: SolverState,
+    ) -> tuple[LangevinX, LangevinX, LangevinX, _ErrorEstimate]:
+        raise NotImplementedError
```
Does this need to be a static method? Is there a reason we must demand that this method be parameterised only at the type level (i.e. via new subclasses) and not at the instance level (i.e. via attributes on `self`)?

andyElking replied:
Fair point, I guess there's no reason not to allow that.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+        solver_state: SolverState,
+        made_jump: BoolScalarLike,
+    ) -> tuple[LangevinTuple, _ErrorEstimate, DenseInfo, SolverState, RESULTS]:
+        del made_jump, args
```
I believe some of these classes are FSAL?

If so then `del made_jump` here is wrong. In the event of a jump then the cached evaluation must be discarded, as we'll have a new value after the jump.

andyElking replied:
Of course, I should have thought of that!

patrick-kidger replied:
Probably you should use `eqxi.unvmap_any(made_jump)` as the predicate. That way we'll evaluate if any of the batch elements require it. (This is also what `AbstractRungeKutta` does.)

The batch elements that don't need this just get the value recomputed pointlessly, of course, but there's ntohing we can do about that. We will at least avoid turning the `cond` into a `select`, for those cases when there's no jump at all!

andyElking replied:
Oh, I'm a true dummy sometimes 😅

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+        # so we use `jnp.isclose` for comparison
+        cond = jnp.isclose(h_prev, h, rtol=1e-10, atol=1e-12)
+        coeffs = lax.cond(
+            cond,
```
Optimization: make this `eqxi.unvmap_all(cond)`.

andyElking replied:
Makes sense.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+
+    term_structure = MultiTerm[tuple[LangevinDriftTerm, LangevinDiffusionTerm]]
+    interpolation_cls = LocalLinearInterpolation
+    taylor_threshold: RealScalarLike = eqx.field(static=True)
```
(Also it should probably be an `AbstractVar` rather than a field? Unusual to need to mandate that something be a field on an abstract class, normally we only care that we can do lookup by name.)

andyElking replied:
That makes sense. I'm still learning the conventions 😅

patrick-kidger commented on diffrax/_solver/quicsort.py:
```diff
@@ -0,0 +1,245 @@
+        self.dtype = jnp.result_type(*all_leaves)
+
+
+class QUICSORT(AbstractLangevinSRK[_QUICSORTCoeffs, None]):
```
(similar comments on this class as on align)

patrick-kidger commented on diffrax/_solver/should.py:
```diff
@@ -0,0 +1,222 @@
+        self.dtype = jnp.result_type(*all_leaves)
+
+
+class ShOULD(AbstractLangevinSRK[_ShOULDCoeffs, None]):
```
(similar comments here too)

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    $d \mathbf{x}_t = \mathbf{v}_t dt$
+
+    $d \mathbf{v}_t = - \gamma \mathbf{v}_t dt - u
+    \nabla f( \mathbf{x}_t ) dt + \sqrt{2 \gamma u} d W_t.$
```
(similar comments on the LaTeX)

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+        return self.control.evaluate(t0, t1, **kwargs)
+
+    def prod(self, vf: LangevinX, control: LangevinX) -> LangevinTuple:
+        dv = vf
```
Nit: this is the wrong name. Vector fields aren't `d`-anything! They're not infinitesimals.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+class LangevinDriftTerm(AbstractTerm):
+    gamma: LangevinX
+    u: LangevinX
+    grad_f: Callable[[LangevinX], LangevinX]
```
Should this accept `args`?

patrick-kidger replied:
(Should it accept a time as well? That one is probably more suspect... I'm not sure what the theory looks like there.)

andyElking replied:
- Args: Technically of course it could, but I would opt for no. That would create a confusion between the Langevin args `(gamma, u, grad_f)` and the args fed into `grad_f` (both for the user and any future developer of Diffrax, who would have to understand the distinction from my comments). So I would prefer the user just makes new `terms` when they want to use a different `grad_f`.
- Time: I think the theory could be modified to allow for time dependence, but all ULD literature out there assumes it is time-homogeneous. But if someone suddenly develops theory for time-dependent ULD, we can add it then.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+
+
+def make_langevin_term(
+    gamma: PyTree[ArrayLike],
+    u: PyTree[ArrayLike],
```
Should these by pytrees with the same structure?

andyElking replied:
Originally yes, but given that you told me to go with the first option (regarding broadcasting these to the structure and shape of `x0`), this is no longer the case, so they needn't have the same structure. In fact a common use-case is that `u` is a scalar, and `gamma` has the same shape&structure as `x0`.

patrick-kidger replied:
Got it!

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+def make_langevin_term(
+    gamma: PyTree[ArrayLike],
+    u: PyTree[ArrayLike],
+    grad_f: Callable,
```
with what signature?

patrick-kidger commented on docs/api/solvers/sde_solvers.md:
```diff
@@ -113,3 +113,40 @@ These are reversible in the same way as when applied to ODEs. [See here.](./ode_
+
+::: diffrax.ShOULD
+    selection:
+        members: false
+
+::: diffrax.QUICSORT
+    selection:
+        members: false
```
For completeness I think I'd prefer to include `__init__` in both of these. (Even though it's the same for all of them.)

patrick-kidger commented on examples/langevin_example.ipynb:
The commentary between the two of you notwithstanding, I really like how short this is! Perfectly captures how to use this API, very concisely.

(If you want to add more examples then we can do the same thing for Langevin as we've done for neural differential equations, which is to write several examples and then group them in the docs. Although if they can all be this short then that would be perfect... !)

andyElking replied:
Great!

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
+        return jtu.tree_map(lambda _vf: control * _vf, vf)
+
+
+def make_langevin_term(
```
Btw, we're calling all of this just 'Langevin', but for completeness our terminology throughout should really be 'underdamped Langevin', I think?

andyElking replied:
I changed it here, although I think having all class names spell out `UnderdampedLangevin` would be a bit much. I do make it clear in the docstrings though. WDYT?

patrick-kidger replied:
Mostly I'm thinking that if in the future, we ever wanted to add similar special support for e.g. overdamped langevin, then we should make sure things are ambiguous now.

For that reason I'm inclined towards the more-verbose option.

andyElking replied:
Sounds good. I'll change it everywhere then.

andyElking replied:
For some things which are not exposed to the user, is it okay to go with `ULD`? For example `ULDLeaf = Shaped[Array, "*uld"]`? Or should I use `UnderdampedLangevin` everywhere?

andyElking replied:
I made the change. I do think it might make the code more readable if we shortened at least the type aliases (like `UnderdampedLangevinX`) which can appear many times in fucntion signatures. Maybe `ULDX`? Or is that not sufficiently clear? I am happy to leave it as `UnderdampedLangevinX`, though.

patrick-kidger commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+# one can compute the SRK coefficients more precisely for a small h.
+
+
+class AbstractLangevinSRK(
```
The solvers of this type are all due to James, right? If so then I nominate that we call these `AbstractFosterLangevinSRK` instead :)

We tend to name solvers after the person who came up with them rather than the problem they solve! :)

andyElking replied:
Yep, sound good :)))

andyElking commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+
+    $$d x_t = v_t dt$$
+
+    $$d v_t = - gamma v_t dt - u ∇f(x_t) dt + (2gammau)^(1/2) dW_t$$
```
Sorry, this one slipped by me entirely. Now I tried to really make the ULD equations correctly formatted everywhere.

andyElking commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+        out = self._tay_coeffs_single(c)
+
+        def check_shape(coeff_leaf):
+            permitted_shapes = [c.shape + (3, 6), c.shape + (1, 6), c.shape + (6,)]
+            assert (
+                coeff_leaf.shape in permitted_shapes
+            ), f"leaf shape: {coeff_leaf.shape}, c shape: {c.shape}"
```
This was actually entirely superfluous. The point is that `QUICSORT` uses trailing dims of `(3,6)` or `(1,6)`, whereas `ALIGN` and `ShOULD` use only `(6,)`. But all of them ensure the shape is correct on their own, so it's fine.

andyElking commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+    def _eval_taylor(h, tay_coeffs: _Coeffs) -> _Coeffs:
+        # Multiplies the pre-computed Taylor coefficients by powers of h.
+        dtype = tay_coeffs.dtype
+        h_powers = jnp.power(h, jnp.arange(0, 6, dtype=h.dtype)).astype(dtype)
+        return jtu.tree_map(
+            lambda tay_leaf: jnp.tensordot(tay_leaf, h_powers, axes=1), tay_coeffs
+        )
```
Yes, that's probably better. Thanks!

andyElking commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+            # Depending on the size of h*gamma choose whether the Taylor expansion or
+            # direct computation is more accurate.
+            cond = h * c < self.taylor_threshold  # c is a leaf of gamma
+            if jnp.ndim(c) == 0:
```
Yep good point.

andyElking commented on diffrax/_solver/langevin_srk.py:
```diff
@@ -0,0 +1,320 @@
+                # otherwise, compute both tay_out and direct_out and select the
+                # correct one for each dimension
+                return lax.cond(
+                    eqxi.unvmap_all(cond), lambda _: tay_out, select_tay_or_direct, None
```
Wasn't aware I could do that. Thanks!

andyElking commented:
Hi Patrick,
I made all the fixes you recommended. I went through the code again today and I think I don't see any more issues anywhere (but you'll probably still find some 😅).

I left a few conversations unresolved because I still need some related guidance.

andyElking commented on diffrax/_term.py:
```diff
@@ -776,3 +785,180 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    u: PyTree[ArrayLike]
+    grad_f: Callable[[UnderdampedLangevinX], UnderdampedLangevinX]
+
+    def vf(
+        self, t: RealScalarLike, y: UnderdampedLangevinTuple, args: Args
+    ) -> UnderdampedLangevinTuple:
+        x, v = y
+        # gamma, u and v can all have different pytree structures, we only know that
+        # gamma and u are prefixes of v (which is the same as x)
+        gamma = _broadcast_pytree(self.gamma, v)
+        u = _broadcast_pytree(self.u, v)
+
+        f_x = self.grad_f(x)
+        vf_x = v
+
+        def fun(_gamma, _u, _v, _f_x):
+            return -_gamma * _v - _u * _f_x
+
+        vf_v = jtu.tree_map(fun, gamma, u, v, f_x)
+        vf_y = (vf_x, vf_v)
+        return vf_y
```
Because we went with the first option regarding when to broadcast `gamma` and `u` to the shape of `x` (as discussed [here](https://github.com/patrick-kidger/diffrax/pull/453#discussion_r1724906904)), we must perform `_broadcast_pytree(gamma)` (and `u`) each time we call `UnderdampedLangevin____Term.vf`. This isn't used in the ULD-specific solvers, but generic solvers call this all the time. I assume XLA optimises this, but do you think this would negatively affect compile time?

Maybe I could do the broadcast as part of the multiplication, but that might make the code a bit uglier and I'm not sure if it would significantly improve the situation.

What's your read on it?

patrick-kidger replied:
I think this looks good to me! I think these operations should all be fairly negligible for compile time.

I'm really happy with this change, it looks a lot cleaner to me.

andyElking replied:
Great, that's what I was hoping for :)

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+
+
+_ErrorEstimate = TypeVar("_ErrorEstimate", None, UnderdampedLangevinTuple)
+ULDArgs = tuple[
```
Should this be `UDL` or `ULD`?

That aside, personally I prefer to use the more-verbose approach to things, i.e. `UnderdampedLangevinArgs`. I assume everyone is using an editor with autocomplete these days.

andyElking replied:
It's ULD for Underdamped Langevin Diffusion. But yes, I can go with the longer version.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+def _broadcast_pytree(source, target_tree):
+    # Broadcasts the source PyTree to the shape and PyTree structure of
+    # target_tree_shape. Requires that source is a prefix tree of target_tree
+    # This is used to case gamma and u to the shape of x0 and v0
```
'case' -> 'broadcast'?

andyElking replied:
I think I wanted to write "cast" but made a typo

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+
+
+class AbstractFosterLangevinSRK(
+    AbstractStratonovichSolver[SolverState],
+    AbstractItoSolver[SolverState],
```
This shouldn't be both of these!

andyElking replied:
Oh yes, fair enough

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+        tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, gamma, tay_coeffs)
+        outer = jtu.tree_structure(gamma)
+        assert inner is not sentinel, "inner tree structure not set"
+        coeffs_with_tree = jtu.tree_transpose(outer, inner, tree_with_coeffs)  # type: ignore
```
nit: can you make this `# pyright: ignore[someErrorCode]`?

We only use pyright so I prefer to use the more specific annotation.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+            def select_tay_or_direct():
+                if jnp.ndim(c) == 0:
+                    direct_out = self._directly_compute_coeffs_leaf(h, c)
+                else:
+                    fun = lambda _c: self._directly_compute_coeffs_leaf(h, _c)
+                    direct_out = vmap(fun)(c)
```
I'd add an `assert jnp.ndim(c) == 1` here, which I assume is what is going on.

andyElking replied:
It's very good you wrote this here, made me realise this vmap is actually redundant. Originally LangevinSRK could only handle rank 0 and 1 tensors, so `self._directly_compute_coeffs_leaf` could only take a scalar `_c`. But now that it can handle arbitrary shapes, I can get rid of this.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+    ) -> tuple[
+        UnderdampedLangevinX, UnderdampedLangevinX, UnderdampedLangevinX, _ErrorEstimate
+    ]:
+        r"""This method specifies how to compute a single step of the ULD SRK
```
`ULD` -> something more informative.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+
+        # compute the Brownian increment and space-time(-time) Levy area
+        levy = diffusion.contr(t0, t1, use_levy=True)
+        assert isinstance(levy, self.minimal_levy_area), (
+            f"The Brownian motion must have"
+            f" `levy_area={self.minimal_levy_area.__name__}`"
+        )
```
I think this should be a `ValueError` rather than an assertion, as it is possible to hit this through user error.

andyElking replied:
Good point.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+        )
+
+        x0, v0 = y0
+        prev_f = lax.cond(
+            eqxi.unvmap_any(made_jump), lambda: grad_f(x0), lambda: st.prev_f
+        )
```
Even better as an optimization here:
```python
if made_jump is False:
    prev_f = st.prev_f
else:
    prev_f = lax.cond(
        eqxi.unvmap_any(made_jump), lambda: grad_f(x0), lambda: st.prev_f
    )
```

This is possible because `made_jump` will frequently be a literal Python `False` (the diffeqsolve machinery tries to set it to this if possible, and indeed most solves don't have jumps). And then when that is the case, we can avoid tracing `grad_f` here at all -- which even if the branch is unconditionally falsey will still be traced by JAX, and will increase compile times as a result.

(This is one of those of things I've learnt to look out for -- minimising the number of times a user function is traced by JAX is an important way to reduce compile times, which can otherwise be fairly long.)

patrick-kidger replied:
Side note we have a `static_select` helper function that does something similar; maybe we should introduce a `static_cond` one too?

Up to you.

andyElking replied:
Given this is a one-time use I suppose there is no need for now.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -0,0 +1,410 @@
+    the friction and the inertia of the system.
+
+    Solvers which inherit from this class include [`diffrax.ALIGN`][],
+    [`diffrax.ShOULD`][], and [`diffrax.QUIC_SORT`][].
```
Nit: spurious underscore in `QUICSORT`.

(Try building the documentation with `mkdocs serve` -- it should probably detect this and error out.)

andyElking replied:
This docstring isn't included in the docs, so it didn't throw an error. But thanks for spotting it.

patrick-kidger commented on diffrax/_solver/quicsort.py:
```diff
@@ -0,0 +1,250 @@
+    ) -> _QUICSORTCoeffs:
+        del self
+        # compute the coefficients directly (as opposed to via Taylor expansion)
+        original_shape = c.shape
```
`jnp.shape(c)`? I don't know if normal Python values can infiltrate this far, but just in case.

patrick-kidger commented on diffrax/_solver/quicsort.py:
```diff
@@ -0,0 +1,250 @@
+        uh = (u**ω * h).ω
+        v_tilde = (v0**ω + rho**ω * (hh**ω + 6 * kk**ω)).ω
+
+        x1 = (x0**ω + a_l**ω * v_tilde**ω + b_l**ω * rho_w_k**ω).ω
+        f1uh = (f(x1) ** ω * uh**ω).ω
+
+        x2 = (
+            x0**ω + a_r**ω * v_tilde**ω + b_r**ω * rho_w_k**ω - a_third**ω * f1uh**ω
+        ).ω
+        f2uh = (f(x2) ** ω * uh**ω).ω
```
Let's not do this now -- leave it for a follow-up PR if you have the energy (+we can then more easily verify correctness) -- but this would be better done as a little length-2 scan, so that we only need to compile `f` once.

See how e.g. `_step_size_controller/adaptive.py::_select_initial_step` uses `eqxi.scan_trick`.

andyElking replied:
Done! It's a very clever trick indeed 😊.

Just how significant are the speedups from this? I imagine a more complicated piece of code (so `eqxi.scan_trick` in this case) brings a computational overhead of its own, so this is surely a tradeoff. Sometimes it's really not obvious to me if it is worth having significantly more convoluted code to achieve some speedup. How do you decide? Do you benchmark these things?

patrick-kidger replied:
The decreased readability is indeed substantial. It's unfortunate.

Generally this is worth doing for anything involving the user-provided vector. These sometimes get very large indeed; every extra trace can increase compile time by several minutes.

andyElking replied:
Oh, so the point is that it's user-supplied, so the worst case scenario can be arbitrarily expensive, whereas we know that the overhead of the code is not that big. Makes sense, thanks!

patrick-kidger commented on diffrax/_solver/quicsort.py:
```diff
@@ -0,0 +1,250 @@
+        ).ω
+        v_out = (v_out_tilde**ω - rho**ω * (hh**ω - 6 * kk**ω)).ω
+
+        # this method is not FSAL, but for compatibility with the base class we set
+        f_fsal = prev_f
```
This seems dodgy. If I understand correctly then this gets piped all the way back around to `prev_f` in the entrypoint to this function, where it is unused, and is therefore technically fine, but even so...

I'd suggest setting `f_fsal = None` here and tweaking the shape-checking code to allow for that. Also, `del prev_f` at the start of this function as an assertion that it is not used.

andyElking replied:
Yeah, that's fair.

I did have to add an `_is_fsal` class var, so that I set `solver_state.prev_f` to the right value in `AbstractFosterLangevinSRK.init` (otherwise I get `body_fun must have the same input and output structure`).

patrick-kidger replied:
I think that's fine!

patrick-kidger commented on diffrax/_solver/should.py:
```diff
@@ -0,0 +1,239 @@
+            + coeffs.a_half**ω * v1**ω
+            + coeffs.b_half**ω * (-(uh**ω) * f0**ω + rho_w_k**ω)
+        ).ω
+        f1 = f(x1)
+
+        chh_hh_plus_ckk_kk = (coeffs.chh**ω * hh**ω + coeffs.ckk**ω * kk**ω).ω
+
+        x_out = (
+            x0**ω
+            + coeffs.a1**ω * v0**ω
+            - uh**ω * coeffs.b1**ω * (1 / 3 * f0**ω + 2 / 3 * f1**ω)
+            + rho**ω * (coeffs.b1**ω * w**ω + chh_hh_plus_ckk_kk**ω)
+        ).ω
+        f_out = f(x_out)
```
Same comment, in a follow-up we could consider switching this to use a scan trick.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,180 @@ def _to_vjp(_y, _diff_args, _diff_term):
+UnderdampedLangevinTuple = tuple[UnderdampedLangevinX, UnderdampedLangevinX]
+
+
+def _broadcast_pytree(source, target_tree):
```
This function has appeared in two different spots now -- let's factor it out into `_misc.py`.

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -776,3 +785,180 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    # as x and v. We only require that they are prefixes of x and v, in a way that
+    # allows us to call _broadcast_pytree(gamma, x) and _broadcast_pytree(u, x).
+
+    drift = UnderdampedLangevinDriftTerm(gamma, u, grad_f)
+    diffusion = UnderdampedLangevinDiffusionTerm(gamma, u, bm)
+    return MultiTerm(drift, diffusion)
```
Look how far this has been minimised! That's awesome.
Basically, we've successfully pushed things into the abstractions that Diffrax expects.

Do you think it'd be worth removing this function altogether, and then exposing `UnderdampedLangevin{Drift,Diffusion}Term` directly? I don't have an immediate strong opinion.

andyElking replied:
True, this really is a nice way of doing it, thanks for the suggestions!

We could, but then we'd have to add more checks that the drift and diffusion have the same arguments. And I don't think it would make the life of users any easier. If they want ULD-like terms, but with mismatched arguments, they can write them themselves, but they'll have to use generic solvers anyway.

In addition, the Langevin example advertises convergence to the well-known stationary distribution, which might not be the case if the arguments are not the same in both terms, so some less-informed users could be confused by this.

patrick-kidger replied:
> We could, but then we'd have to add more checks that the drift and diffusion have the same arguments. And I don't think it would make the life of users any easier. If they want ULD-like terms, but with mismatched arguments, they can write them themselves, but they'll have to use generic solvers anyway.

It's more a matter of making it easy to branch:
```python
drift = UnderdampedLanveginDriftTerm(...)
if foo:
    diffusion = UnderdampedLangevinDiffusionTerm(...)
    solver = SomeLangevinSolver(...)
else:
    diffusion = ControlTerm(...)
    solver = Euler()
```
this kind of thing gets harder the more wrappers that are around. I think having the terms be separate is also conceptually simpler -- they look a lot more like all the other `*Term`s we have floating around.

My main concern with splitting it out is making sure that everything operates 'safely'. For example, if someone were to do:
```python
drift = UnderdampedLangevinDriftTerm(gamma1, ...)
diffusion = UnderdampedLangevinDiffusionTerm(gamma2, ...)
solver = ... # one of the Langevin solvers
```
then things could go awry.

I suppose we could handle this edge case by having `gamma1 != gamma2` change the `result` output of `solver.step`, so that we later raise an error.

On balance I think splitting this out, and setting `result`, is probably my preferred approach, but it's not a strong opinion.

andyElking replied:
That sounds reasonable. I can make the change.

The point where drift args and diffusion args are most easily compared is in `solver.init`, but I can

1. pump that info through `solver_state` into `step` where it can then affect `result`,
2. compare them each time `step` is called (which I think would incur unnecessary computation), or
3. simply use an `eqx.error_if` in `_get_args_from_terms` inside `solver.init`.

What's your preferred option?

andyElking replied:
I also realised that they can in fact do that anyway, by unpacking the `MultiTerm`. But yes in this case it's better to formalise it, add the checks and give users full freedom.

patrick-kidger replied:
On reflection I think probably option 3 is best, actually!

andyElking replied:
Fair enough, will do.

patrick-kidger commented on docs/api/solvers/sde_solvers.md:
```diff
@@ -113,3 +113,43 @@ These are reversible in the same way as when applied to ODEs. [See here.](./ode_
+These solvers are specifically designed for the Underdamped Langevin diffusion (ULD),
+which takes the form 
+
+\begin{align*}
+    \mathrm{d} x(t) &= v(t) \, \mathrm{d}t \\
+    \mathrm{d} v(t) &= - \gamma \, v(t) \, \mathrm{d}t - u \,
+    \nabla \! f( x(t) ) \, \mathrm{d}t + \sqrt{2 \gamma u} \, \mathrm{d} w(t),
+\end{align*}
```
Does this get rendered correctly in the docs? It might need some surrounding `$`?

andyElking replied:
It does for me. Do you think I should still try to use `$`?

patrick-kidger replied:
As long as it doe! I've not tried myself.

patrick-kidger commented on docs/devdocs/srk_example.ipynb:
I think changes here have snuck in from your other branch, as `JumpStepWrapper` does not exist here yet.

andyElking replied:
Oops

patrick-kidger commented on examples/underdamped_langevin_example.ipynb:
Similar comment here, does `\begin{align*}` render correctly in the docs?

patrick-kidger replied:
(That aside, this example is super short, which is great! Thank you for that.)

patrick-kidger reviewed (commented):
Okay, this basically LGTM!

I've gone through and left what I think is a final round of nits, but I think they should all be super minor.

andyElking commented:
Quick question @patrick-kidger: 

Should I make `AbstractFosterLangevinSRK` public and add it under Abstract Solvers in the docs? I haven't done it so far, because, unlike RK and SRK, where the user just has to specify a tableau, making a custom child of `AbstractFosterLangevinSRK` is much more involved, so I doubt users who are just using a packaged version of Diffrax would do that. WDYT?

patrick-kidger commented:
I think it probably should be public + in the abstract solvers. I agree that writing your own here is *incredibly* niche, but I think it's useful for inquisitive users to be able to poke at such things.

andyElking commented on diffrax/_term.py:
```diff
@@ -776,3 +785,266 @@ def _to_vjp(_y, _diff_args, _diff_term):
+UnderdampedLangevinTuple = tuple[UnderdampedLangevinX, UnderdampedLangevinX]
+
+
+class UnderdampedLangevinStructureError(Exception):
```
I would like your opinion on this Patrick. I realised that if `gamma` or `u` are indeed shaped incorrectly, then Diffrax throws the following:
```python
ValueError: `terms` must be a PyTree of `AbstractTerms` (such as `ODETerm`), with structure diffrax._term.MultiTerm[tuple[diffrax._term.AbstractTerm[typing.Any, typing.Union[jaxtyping.Float[Array, ''], jaxtyping.Float[ndarray, ''], numpy.number, float, jaxtyping.Int[Array, ''], jaxtyping.Int[ndarray, ''], int]], diffrax._term.AbstractTerm[typing.Any, diffrax._custom_types.AbstractSpaceTimeLevyArea]]]
```

This comes from `_integrate._term_compatible` and is rather unhelpful. But if we want a different error message, then we must throw an exception that is not a `ValueError`, because those are caught by `_term_compatible`. Do you think this solution is appropriate, or do you suggest something else?

Also because of this I kept `_broadcast_pytree` in here, and then the function `broadcast_underdamped_langevin_arg` defined below gets imported by `foster_langevin_srk.py`. Is that okay?

patrick-kidger replied:
Switching that to a different error type sound reasonable. .Although it shouldn't matter too much, the original error should still be visible in the stack trace above.

Rather than creating a new error type though, I think let's just make this one a `RuntimeError` for simplicity.

The import seems fine to me.

andyElking replied:
Good, I'll swap it for a `RuntimeError` then.

andyElking commented:
So I added the check that drift and diffusion have the same arguments in `AbstractFosterLangevinSRK.init` and also added a short test for this in `test_underdamped_langevin.py`.

I will now do the scan trick.

andyElking commented:
I think I now addressed everything you suggested, including the scan trick in both `QUICSORT` and `ShOULD`. I'll do another quick check and then I pass it back to you.

andyElking commented:
I went through the code again and now I think I addressed everything you mentioned. Also please take a look at this comment [here](https://github.com/patrick-kidger/diffrax/pull/453#discussion_r1740114406) and let me know if I should revert it to how it was before. Other than that there are no major changes.

andyElking commented:
Sorry, I left a tiny problem in the test I added, I fixed it now.

patrick-kidger commented on diffrax/_solver/foster_langevin_srk.py:
```diff
@@ -384,7 +416,13 @@ def check_shapes_dtypes(_x, _v, _f, _x0):
+            _f_fsal = f_fsal
+        else:
+            assert f_fsal is None
+            _f_fsal = x0
```
Assigning this anyway is a little dangerous. Perhaps instead:
```python
def check_shapes_dtypes(arg, *args):
    for x in args:
        assert arg.shape == x.shape
        assert arg.dtype == x.dtype

if self._is_fsal:
    jtu.tree_map(check_shapes_dtypes, x_out, v_out, x0, f_fsal)
else:
    assert f_fsal is None
    jtu.tree_map(check_shapes_dtypes, x_out, v_out, x0)
```
?

andyElking replied:
That's a very elegant solution!

patrick-kidger commented on docs/api/terms.md:
```diff
@@ -92,4 +92,40 @@ Some example term structures include:
+        members:
+            - __init__
+
 ::: diffrax.make_underdamped_langevin_term
```
Shall we remove this, then?

andyElking replied:
I moved the function to `test.helpers` because I use it several times there, but I removed it everywhere else.

patrick-kidger commented:
Aaaaand... merged! 🎉

andyElking commented:
Thanks for bearing with me and taking your time Patrick!! I really appreciate it!
